### ISSUE TITLE
Vulkan

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#102]
While it probably won't make the game any faster since it's not API bound, I think Vulkan does offer some advantages:
- better tooling (there is a GPU profiler from AMD that only works with Vulkan)
- better platform support: with MoltenVK there is apple support. And Vulkan also supports mobile GPUs, which at the moment would require OpenGL ES
- better error detection with validation layers. OpenGL is really inconsistent with errors and warnings, which appear to be driver dependant.
- Vulkan does provide more consistent performance. For example on my intel laptop the OpenGL driver creates a copy of a given buffer when doing glBufferSubData.
- Vulkan is probably more future-proof at this point.

But all that comes at a heavy development cost.

Prerequisites:
- #101 would make porting easier

**Comment by jujubeja** [comment]
https://github.com/ashpil/moonshine A general purpose ray traced renderer built with Zig + Vulkan

**Comment by Argmaster** [comment]
There is also https://vulkan.gpuinfo.org/ page which is very handy for querying for how widely are different Vulkan features available.

**Comment by ricknijhuis** [comment]
Question, I did see the use of glad for loading the vulkan function pointers and providing the headers. Was the following project ever considered https://github.com/Snektron/vulkan-zig? to do that instead? It has some large advantages over a c library like glad. 
- Provides a type safe zig interface (packed structs instead of flags, enums as handles instead of opaque ptr's)
- Provides an error set per function making it very clear what errors can be thrown by what functions
- Provides some convenience enumerate functions that return an allocated slice.

I have used a custom fork of this library that worked very well for my cases. It removed the proxies it provided, I preferred keeping the vulkan code more to the original c api, functions can be called like try vk.destroyInstance() instead of when using a proxy call like instance.destroyInstance().

It would mean replacing glad third party with vulkan-zig as third party. If you don't like having another lib you could just generate the bindings once and include it directly in the project source. (If going this route you can always fork the generator so you are in control over the code.) 

If no objections I could provide a pr updating the current vulkan code. This seems like a great project and I would like to contribute to it. 

If this is the wrong place to ask these kind of questions please let me know.

**Comment by Argmaster** [comment]
C libraries have a great advantage over zig counterparts - they have a stable API. Zig changes dynamically and Quantum tries his best to keep up with this, but he doesn't have the time to understand and maintain someone's else's codebase if they happen to give up doing so.
Additionally this adds cost of having to understand abstraction someone invented on top of Vulkan. Using plain C API means you can refer to majority of existing online resources. Further you go from it, the harder it will become to translate one into the other.
That however is only my opinion.

**Comment by ricknijhuis** [comment]
I do get the stable API part though in this case I'm note sure it really applies here. The language constructs used in the library are quite basic. I don't see those change any time soon. 

The abstraction as well is absolutely minimal and basically the same as with glad. The library generates it's bindings from the vk.xml same as with glad, the only difference here is glad does it in c and vulkan-zig in zig, making use of some zig constructs to make it more type safe. if you would target for example vk 1.3 you would just need to generate the bindings once and they should work forever. 

Here a sample of the generated code for createInstance:
```zig  
pub const CreateInstanceError = error{
    OutOfHostMemory,
    OutOfDeviceMemory,
    InitializationFailed,
    LayerNotPresent,
    ExtensionNotPresent,
    IncompatibleDriver,
    ValidationFailed,
    Unknown,
};
pub fn createInstance(
    p_create_info: *const InstanceCreateInfo,
    p_allocator: ?*const AllocationCallbacks,
) CreateInstanceError!Instance {
    var out_instance: Instance = undefined;
    const result = base_dispatch.vkCreateInstance.?(
        p_create_info,
        p_allocator,
        &out_instance,
    );
    switch (result) {
        Result.success => {},
        Result.error_out_of_host_memory => return error.OutOfHostMemory,
        Result.error_out_of_device_memory => return error.OutOfDeviceMemory,
        Result.error_initialization_failed => return error.InitializationFailed,
        Result.error_layer_not_present => return error.LayerNotPresent,
        Result.error_extension_not_present => return error.ExtensionNotPresent,
        Result.error_incompatible_driver => return error.IncompatibleDriver,
        Result.error_unknown => return error.Unknown,
        Result.error_validation_failed => return error.ValidationFailed,
        else => return error.Unknown,
    }
    return out_instance;
}
```
It maps almost 1:1 to the normal vulkan functions instead of using an out parameter and return error code it returns error or result. I don't think any of these used zig language constructs will change any time soon?

**Comment by IntegratedQuantum** [comment]
> and vulkan-zig in zig, making use of some zig constructs to make it more type safe.

This part is exactly the problem. All these wrapper libraries make small changes throughout the API making it more difficult to use with the official documentation and tutorials. The renaming of things (like c.vkCreateInstance to vk.createInstance) alone is already making this more difficult for no good reason.

**Comment by linux-universe** [comment]
I wonder who will be faster, mojang or you guys :)

**Comment by Samueru-sama** [comment]
I don't think it is a good idea to replace opengl for vulkan, you will run into a lot of bugs in vulkan. 

For example GTK4 now has a vulkan renderer, it is full of bugs and people often have to set `GSK_RENDERER=gl` to fix those issues.

The vulkan renderer was also totall broken with intel gpus (even the latest gpus) and mesa had to [fix it with a workaround](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38373)

At pkgforge-dev we also experimented with shipping a trimmed down version of mesa with only zink as the opengl driver, but that went nowhere because on some systems you have to set [`LIBGL_KOPPER_DRI2=1` for zink to work](https://gitlab.freedesktop.org/mesa/mesa/-/work_items/13929). I also had to [bring back nouveau GL](https://github.com/pkgforge-dev/Anylinux-AppImages/issues/527#issuecomment-4301538906) because the performance of zink is bad on old nvidia gpus to the point that it is noticeable on simple GUI apps.

---

Vulkan only makes sense for the performance gains it offers, yuzu (and now eden) have been focusing on it lately, but have had to keep opengl around because some games on only work with opengl, specially on intel...

Like I once had to make [this report](https://gitlab.freedesktop.org/mesa/mesa/-/work_items/12345) to mesa to give you an idea xd

**Comment by IntegratedQuantum** [comment]
The GTK4 example looks to me like it's an implementation issue on their side not synchronizing things correctly.

I do have trust in Vulkan drivers, because e.g. Steam uses exclusively Vulkan to run Windows games, so at the very least it is possible to achieve wide support. And in general Vulkan is the future, and we have a much better chance to see bug fixes for vulkan drivers than for OpenGL drivers (and I did have my trouble with that multiple times already, to the point where they don't even support OpenGL 4.6 core features #1381, and these problems will only grow as more and more applications move over to Vulkan).

> Vulkan only makes sense for the performance gains it offers

There are no performance gains from Vulkan to be had for Cubyz.
I want to use it for its wider support (macos does not support modern OpenGL at all) and better support for extensions (e.g. #817 has more device coverage on Vulkan).

**Comment by RanPix** [comment]
> There are no performance gains from Vulkan to be had for Cubyz.
I want to use it for its wider support (macos does not support modern OpenGL at all) and better support for extensions (e.g. https://github.com/PixelGuys/Cubyz/issues/817 has more device coverage on Vulkan).

also the 4gb ssbo sizes

**Comment by tillpp** [comment]
Main reason for me liking vulkan:
u can split big shader files into smaller readable files. 
(you couldnt do that without extension in opengl and using spirv in gl wasn't  often supported)

**Comment by Samueru-sama** [comment]
> I do have trust in Vulkan drivers

Alright wish you the best luck.

I do not think vulkan drivers will improve in the future (specially anything that uses hasvk) and looks like the intel vulkan drivers will also have another split with the new jay shader compiler.

On the other hand AMD side of mesa has been good, with great vulkan support all the way down to the HD7000 series gpus. This of course was possible thanks to all the people Valve has had contributing to.

**Comment by IntegratedQuantum** [comment]
Looks like we can easily support Vulkan 1.3
<img width="477" height="1056" alt="Image" src="https://github.com/user-attachments/assets/8f79b869-e510-41be-b002-13e8b8c1826e" />
